### DISCUSSION TITLE
How librespot works?

### DISCUSSION THREAD

**Discussion by carrieroent** [librespot-org/librespot#1132] [General]
Hello, I would like answers to the following questions :)

What APIs does liberaspot use? Spotify internal, private API only?

How technically we able to use the Spotify private API? Did we take the keys for the private API from official clients? How did it do from obfuscated, binary?

If we got the keys from the official clients, is there any way Spotify can block us? And can do the same trick for any service?

What does reversing private API mean in practice?

Why can't we enable Spotify Free accounts? Can't Spotify private API put ads to streams directly?

Why Spotify doesn't allow third parties to use the 'streaming API'? I rent music them with money or ads, why does it matter to them which client I use?

Slightly off-topic: 

Spotify has public API for metadata, but does any other music streaming service (Deezer, Tidal, Qobuz, Apple Music or others) have public 'streaming API'?

What other than Spotify has something like librespot possible?

**Comment by roderickvd** [comment]
For the benefit of the doubt I will give this one pass. As you have been asking such questions to something like four other projects at the same time, I very much encourage you to tell what you're up to which is worth taking the time of the community.

> What APIs does liberaspot use? Spotify internal, private API only?

Private API -- I already said so in #1130. Please see the source code and don't ask the same question twice.

> How technically we able to use the Spotify private API? Did we take the keys for the private API from official clients? How did it do from obfuscated, binary?

It was reverse engineered by analysing the network traffic.

> If we got the keys from the official clients, is there any way Spotify can block us? And can do the same trick for any service?

Yes, and they have blocked most other projects that depend on the old `libspotify` based API: https://developer.spotify.com/community/news/2022/04/12/libspotify-sunset/
We seem to have been whitelisted.

> What does reversing private API mean in practice?

Besides analysing the network traffic, we have extracted protobuf definitions from the binaries.
Please Google it if you don't know what that means, but want to.

> Why can't we enable Spotify Free accounts? Can't Spotify private API put ads to streams directly?

We could but:
1. It takes developer time
2. We maintain the assumption that it's easier for Spotify to tolerate us knowing it's all paying users

This has been discussed ad nauseam many times. Please use the search feature.

> Why Spotify doesn't allow third parties to use the 'streaming API'? I rent music them with money or ads, why does it matter to them which client I use?

Ask them.

> Spotify has public API for metadata, but does any other music streaming service (Deezer, Tidal, Qobuz, Apple Music or others) have public 'streaming API'?

Not that I know of.

> What other than Spotify has something like librespot possible?

Depends on what feature set you mean. Many people use `librespot` for its Connect functionality. As far as I know, there is no other streaming service that has anything like it. I have gotten quite far with Deezer Connect but got stuck on the synchronisation handshake -- have not had development time since, and won't have for a long time.

**Reply by carrieroent** [reply]
> For the benefit of the doubt I will give this one pass. As you have been asking such questions to something like four other projects at the same time, I very much encourage you to tell what you're up to which is worth taking the time of the community.

Thanks, your answer helped a lot! Note that not all the questions I ask them are about the same thing. My purpose is just to know how these softwares works and compare music streaming services 'support' for integration, so I can decide where to spend my time. For example, Tidal's third-party clients are quite varied.

> > What APIs does liberaspot use? Spotify internal, private API only?
> 
> Private API -- I already said so in https://github.com/librespot-org/librespot/issues/1130. Please see the source code and don't ask the same question twice.

Can you link to the points of source code that contain the keys and the URL of the API. I'm a pretty bad programmer and I didn't find it, but I'm interested in testing a couple of things.

>> How technically we able to use the Spotify private API? Did we take the keys for the private API from official clients? How did it do from obfuscated, binary?
> 
> It was reverse engineered by analysing the network traffic.

I understand that by analysing the network traffic you can find out the URL of the API and possibly examine the protocol, but how could the traffic reveal the keys to the API? I'm pretty sure Spotify encrypts the traffic.

>> If we got the keys from the official clients, is there any way Spotify can block us? And can do the same trick for any service?
> 
> Yes, and they have blocked most other projects that depend on the old libspotify based API: https://developer.spotify.com/community/news/2022/04/12/libspotify-sunset/
We seem to have been whitelisted.

They can disable the old API, but if we use the private API that official clients use as well and with the these clients keys, how can they block us? If they close/block the keys, the official clients will also stop working, right?

Unanswered: "And can do the same trick for any service?"

Unasnwered: "Spotify private API put ads to streams directly, right?"

>> Why Spotify doesn't allow third parties to use the 'streaming API'? I rent music them with money or ads, why does it matter to them which client I use?
> 
> Ask them.

I mean what people think is the reason.

>> What other than Spotify has something like librespot possible?
> 
> Depends on what feature set you mean. Many people use librespot for its Connect functionality. As far as I know, there is no other streaming service that has anything like it. I have gotten quite far with Deezer Connect but got stuck on the synchronisation handshake -- have not had development time since, and won't have for a long time.

I mean the streaming features, connect isn't important for me and not many services have it at all. For example, [hifi.rs](https://github.com/iamdb/hifi.rs) seems to work similarly to librespot, right? So with Qobuz something like librespot is possible. Are there others? Is this possible with e.g. Deezer or Tidal?

Note: You don't have to answer these follow-up questions if you don't want to. (I ref to "I will give this one pass.")
